src/Components/ServiceScene/Breakdowns/ByFrameRepeater.tsx

```diff
@@ -63,28 +65,42 @@ export class ByFrameRepeater extends SceneObjectBase<ByFrameRepeaterState> {
     }
   };
 
-  private getSortedSeries(data: PanelData) {
-    const reducer = fieldReducers.get(this.sortBy);
-
-    const fieldCalcs = data.series.map((dataFrame) => ({
-      value: reducer.reduce?.(dataFrame.fields[1], true, true) ?? doStandardCalcs(dataFrame.fields[1], true, true),
-      field: dataFrame,
+  private getSortedSeries = (data: PanelData) => {
+    const reducer = (dataFrame: DataFrame) => {
+      if (this.sortBy === 'changepoint') {
+        return this.changePointsValue(dataFrame);
+      }
+      const fieldReducer = fieldReducers.get(this.sortBy);
+      const value =
+        fieldReducer.reduce?.(dataFrame.fields[1], true, true) ?? doStandardCalcs(dataFrame.fields[1], true, true);
+      return value[this.sortBy] ?? 0;
+    };
+
+    const seriesCalcs = data.series.map((dataFrame) => ({
+      value: reducer(dataFrame),
+      dataFrame: dataFrame,
     }));
 
-    fieldCalcs.sort((a, b) => {
-      // reducerValue will be a Record<ReducerID, number> or an empty object {}
-      if (a.value[this.sortBy] !== undefined && b.value[this.sortBy] !== undefined) {
-        return b.value[this.sortBy] - a.value[this.sortBy];
+    seriesCalcs.sort((a, b) => {
+      if (a.value && b.value) {
+        return b.value - a.value;
       }
       return 0;
     });
 
     if (this.direction === 'asc') {
-      fieldCalcs.reverse();
+      seriesCalcs.reverse();
     }
 
-    return fieldCalcs.map(({ field }) => field);
-  }
+    return seriesCalcs.map(({ dataFrame }) => dataFrame);
+  };
+
+  private changePointsValue = (data: DataFrame) => {
+    const fields = data.fields.filter((f) => f.type === FieldType.number);
+    const values = new Float64Array(fields[0].values);
```

Comment:
If it's still too slow, we could consider sampling these points before feeding them to the detector (something like this?)

```suggestion
    const values = new Float64Array(fields[0].values.filter((_, i) => i % 5 === 0));
```

Will need to experiment to see how it runs though.

Reply:
Sampling definitely makes a huge difference in performance, but also in results. When results are sampled, the order changes, sometimes significantly.

Tab on the left, sampling every 3, and on the right with no sampling (claerly the slowest tab), with data of the last 6 hours:

https://github.com/grafana/explore-logs/assets/1069378/3f66895e-4ab5-4c4d-915b-0e5558d2ace7

I'll see what happens using different sample sizes based on the data set length.

Reply:
I updated this to define the "step" to take samples based on the time series length with a simple rule:
- If the series length is below 100, the step is 1
- Else, step will be taking in multiples of 100

For example:

L = series length
S = step

```
L   | S | Final sample size
1   | 1 | 1
361 | 3 | 121
127 | 1 | 127
324 | 3 | 108
43  | 1 | 3
229 | 2 | 115
3   | 1 | 3
```

With that, I'm getting more consistent sorting results between sampled and non sampled.

https://github.com/grafana/explore-logs/assets/1069378/80647282-70f4-4808-9d92-751e387ca553

Reply:
Nice, that sounds very sensible to me! :tada: